### PR TITLE
fix: further optimize archive workflow listing. Fixes #13601

### LINKED ISSUE #13601: 3.5: Further optimize Archive List API call / DB query
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

This is a follow-up to https://github.com/argoproj/argo-workflows/issues/13295 and https://github.com/argoproj/argo-workflows/pull/13566. We can further improve the query performance, seemingly by another order of magnitude, by adding a union index:

> > ```sql
> > CREATE INDEX argo_archived_workflows_i5 ON argo_archived_workflows (clustername, startedat);
> > ```

This can be added as a migration after [this line in `migration.go`](https://github.com/argoproj/argo-workflows/blob/8065edb34df0e188fd88181b4660c705997e6dcf/persist/sqldb/migrate.go#L260).

We should also test the performance of this index with and without the subquery. If the subquery is no longer necessary, then we can revert #13566 along with this change

#### More context

Quoting myself in https://github.com/argoproj/argo-workflows/pull/13566#pullrequestreview-2301010281:
> Note that this query seems to be further optimizable per https://github.com/argoproj/argo-workflows/issues/13295#issuecomment-2340964248, so we may want to open a new issue to add further optimizations, but this at least fixes the performance issues and gets us to a much more reasonable baseline

Quoting @Danny5487401 from https://github.com/argoproj/argo-workflows/issues/13563#issuecomment-2339660938
> > @agilgur5 three methods i tested:
> > 
> > 1. nice implemention by sub,according to https://github.com/argoproj/argo-workflows/pull/13566
> > 
> > [...]
> >   
> > it drops from 37s 768ms to 1 s 981 ms
> > 
> > 2. force index (argo_archived_workflows_i4)
> > 
> > ```sql
> > SELECT
> >     [...]
> > FROM `argo_archived_workflows` force index (argo_archived_workflows_i4)
> > WHERE [...]
> > ORDER BY `startedat` DESC
> > LIMIT 20;
> > ```
> > 
> > it drops from 37s 768ms to 138 ms
> > 
> > 3. union index (clustername, startedat) to use Backward index scan instead of filesort
> > 
> > ```sql
> > CREATE INDEX argo_archived_workflows_i5 ON argo_archived_workflows (clustername, startedat);
> > ```
> >   
> > it drops from 37s to 140 ms

And then my response:

> It sounds like the `force index` and new index are both faster than the subquery, at least on MySQL 8.0, I wonder if we should use those instead of the subquery. `force index` is potentially not the most resilient, but the new index seems like a "why not?" to me

We should also add a simplified/minimal "performance regression test" to ensure that, well, the performance doesn't regress, per https://github.com/argoproj/argo-workflows/pull/13566#issuecomment-2336734101:

### Version(s)

v3.5, v3.5.10, 729ac17

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
n/a - this is API focused
```

### Logs from the workflow controller

```text
n/a - this is API focused
```

### Logs from in your workflow's wait container

```text
n/a - this is API focused
```

### PR DESCRIPTION
**Pull request by MasonM**
Fixes #13601

### Motivation

Listing archived workflows can be slow if you have a very large number of workflows (100,000+), or the average workflow size is high (100KB), even after the optimizations from #13566 and #13779. This makes some additional optimizations that speed up the queries by ~90% on MySQL and ~50% on PostgreSQL. 

### Modifications
The bottleneck for these queries depended on whether you use MySQL or PostgreSQL, each of which required a different fix. For PostgreSQL, the bottleneck was detoasting overhead, as explained in https://github.com/argoproj/argo-workflows/issues/13601#issuecomment-2420499551. The fix was to use a [common table expression](https://www.postgresql.org/docs/current/queries-with.html) to reduce the amount of times `workflow` needs to be detoasted, as suggested by @kodieg in https://github.com/argoproj/argo-workflows/issues/13601#issuecomment-2421794871. The new query looks like this:
```sql
WITH workflows AS (
  SELECT
    "name",
    "namespace",
    "uid",
    "phase",
    "startedat",
    "finishedat",
    coalesce(workflow->'metadata', '{}') as metadata,
    coalesce(workflow->'status', '{}') as status,
    workflow->'spec'->>'suspend' as suspend
  FROM "argo_archived_workflows"
  WHERE (("clustername" = $1 AND "namespace" = $2 AND "instanceid" = $3))
  ORDER BY "startedat" DESC
  LIMIT 100
) (
  SELECT
    "name",
    "namespace",
    "uid",
    "phase",
    "startedat",
    "finishedat",
    coalesce(metadata->>'labels', '{}') as labels,
    coalesce(metadata->>'annotations', '{}') as annotations,
    coalesce(status->>'progress', '') as progress,
    coalesce(metadata->>'creationTimestamp', '') as creationtimestamp,
    "suspend",
    coalesce(status->>'message', '') as message,
    coalesce(status->>'estimatedDuration', '0') as estimatedduration,
    coalesce(status->>'resourcesDuration', '{}') as resourcesduration
  FROM "workflows"
)
```

For MySQL, the bottleneck was the optimizer inexplicably refusing to use the `argo_archived_workflows_i4` index and instead using the primary key, which is much more expensive. As explained by @Danny5487401  in https://github.com/argoproj/argo-workflows/issues/13563#issuecomment-2339660938, two ways of solving that are using `FORCE INDEX` or adding a union index on `(clustername, startedat)`. Using `FORCE INDEX` is slightly hacky, and adding a new index is needlessly wasteful when we already have `argo_archived_workflows_i4`, so I opted to modify that index to cover `(clustername, startedat)`. The new query looks like this:
```sql
SELECT
  `name`,
  `namespace`,
  `uid`,
  `phase`,
  `startedat`,
  `finishedat`,
  coalesce(workflow->'$.metadata.labels', '{}') as labels,
  coalesce(workflow->'$.metadata.annotations', '{}') as annotations,
  coalesce(workflow->>'$.status.progress', '') as progress,
  coalesce(workflow->>'$.metadata.creationTimestamp', '') as creationtimestamp,
  workflow->>'$.spec.suspend',
  coalesce(workflow->>'$.status.message', '') as message,
  coalesce(workflow->>'$.status.estimatedDuration', '0') as estimatedduration,
  coalesce(workflow->'$.status.resourcesDuration', '{}') as resourcesduration
FROM `argo_archived_workflows`
WHERE ((`clustername` = ?  AND `namespace` = ? AND `instanceid` = ?))
ORDER BY `startedat` DESC
LIMIT 100
```

### Verification

First, I used https://github.com/argoproj/argo-workflows/pull/13715 to generate 100,000 randomized workflows, with https://gist.github.com/MasonM/52932ff6644c3c0ccea9e847780bfd90 as a template:
* PostgreSQL - `time go run ./hack/db fake-archived-workflows  --template "@very-large-workflow.yaml" --rows 100000`
* MySQL - `time go run ./hack/db fake-archived-workflows  --template "@very-large-workflow.yaml" --rows 100000 -d 'mysql:password@tcp/argo'`

Then, I ran `make BenchmarkWorkflowArchive ` once on the `main` branch and once on this branch (with migration applied), and used  [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to compare:
* PostgreSQL
    ```
    $ benchstat postgres_before.txt postgres_after2.txt 
    goos: linux
    goarch: amd64
    pkg: github.com/argoproj/argo-workflows/v3/test/e2e
    cpu: 12th Gen Intel(R) Core(TM) i5-12400
                                                         │ postgres_before.txt │        postgres_after2.txt         │
                                                         │       sec/op        │   sec/op     vs base               │
    WorkflowArchive/ListWorkflows-12                              25.110m ± 1%   9.694m ± 3%  -61.39% (p=0.002 n=6)
    WorkflowArchive/ListWorkflows_with_label_selector-12           26.14m ± 3%   11.06m ± 3%  -57.70% (p=0.002 n=6)
    WorkflowArchive/CountWorkflows-12                              11.84m ± 1%   11.89m ± 2%        ~ (p=0.310 n=6)
    geomean                                                        19.81m        10.84m       -45.26%
    
                                                         │ postgres_before.txt │        postgres_after2.txt         │
                                                         │        B/op         │     B/op      vs base              │
    WorkflowArchive/ListWorkflows-12                              497.2Ki ± 0%   499.2Ki ± 0%  +0.40% (p=0.002 n=6)
    WorkflowArchive/ListWorkflows_with_label_selector-12          504.0Ki ± 0%   504.9Ki ± 0%  +0.18% (p=0.002 n=6)
    WorkflowArchive/CountWorkflows-12                             8.938Ki ± 1%   8.948Ki ± 2%       ~ (p=0.623 n=6)
    geomean                                                       130.8Ki        131.1Ki       +0.23%
    
                                                         │ postgres_before.txt │         postgres_after2.txt         │
                                                         │      allocs/op      │  allocs/op   vs base                │
    WorkflowArchive/ListWorkflows-12                               8.370k ± 0%   8.393k ± 0%  +0.27% (p=0.002 n=6)
    WorkflowArchive/ListWorkflows_with_label_selector-12           8.405k ± 0%   8.428k ± 0%  +0.26% (p=0.002 n=6)
    WorkflowArchive/CountWorkflows-12                               212.0 ± 0%    212.0 ± 0%       ~ (p=1.000 n=6) ¹
    geomean                                                        2.462k        2.466k       +0.18%
    ¹ all samples are equal
    ```
* MySQL
    ``` 
    $ benchstat mysql_before.txt mysql_after.txt 
    goos: linux
    goarch: amd64
    pkg: github.com/argoproj/argo-workflows/v3/test/e2e
    cpu: 12th Gen Intel(R) Core(TM) i5-12400
                                                         │ mysql_before.txt │           mysql_after.txt           │
                                                         │      sec/op      │    sec/op     vs base               │
    WorkflowArchive/ListWorkflows-12                           43.510m ± 4%   1.650m ± 20%  -96.21% (p=0.002 n=6)
    WorkflowArchive/ListWorkflows_with_label_selector-12       69.646m ± 2%   4.513m ± 24%  -93.52% (p=0.002 n=6)
    WorkflowArchive/CountWorkflows-12                           27.37m ± 4%   28.98m ± 24%        ~ (p=0.394 n=6)
    geomean                                                     43.61m        5.998m        -86.25%
    
                                                         │ mysql_before.txt │          mysql_after.txt           │
                                                         │       B/op       │     B/op      vs base              │
    WorkflowArchive/ListWorkflows-12                           477.3Ki ± 0%   457.6Ki ± 0%  -4.13% (p=0.002 n=6)
    WorkflowArchive/ListWorkflows_with_label_selector-12       488.1Ki ± 0%   461.0Ki ± 0%  -5.54% (p=0.002 n=6)
    WorkflowArchive/CountWorkflows-12                          7.937Ki ± 1%   7.944Ki ± 0%       ~ (p=0.461 n=6)
    geomean                                                    122.7Ki        118.8Ki       -3.22%
    
                                                         │ mysql_before.txt │           mysql_after.txt           │
                                                         │    allocs/op     │  allocs/op   vs base                │
    WorkflowArchive/ListWorkflows-12                            8.224k ± 0%   7.705k ± 0%  -6.31% (p=0.002 n=6)
    WorkflowArchive/ListWorkflows_with_label_selector-12        8.277k ± 0%   7.732k ± 0%  -6.58% (p=0.002 n=6)
    WorkflowArchive/CountWorkflows-12                            205.0 ± 0%    205.0 ± 0%       ~ (p=1.000 n=6) ¹
    geomean                                                     2.408k        2.303k       -4.35%
    ¹ all samples are equal
    ```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -39,6 +39,7 @@ E2E_PARALLEL          ?= 20
 E2E_SUITE_TIMEOUT     ?= 15m
 GOTEST                ?= go test -v -p 20
 ALL_BUILD_TAGS        ?= api,cli,cron,executor,examples,corefunctional,functional,plugins
+BENCHMARK_COUNT       ?= 6
```
Comment (MasonM) [inline Makefile:42]: This is so benchstat gives confidence intervals. Without this, it'll show the message `¹ need >= 6 samples for confidence interval at level 0.95`

File: persist/sqldb/migrate.go [hunk 2]
```diff
@@ -269,6 +269,12 @@ func (m migrate) Exec(ctx context.Context) (err error) {
 			noop{},
 			ansiSQLChange(`alter table argo_archived_workflows alter column workflow set data type jsonb using workflow::jsonb`),
 		),
+		// change argo_archived_workflows_i4 index to include clustername so MySQL uses it for listing archived workflows. #13601
+		ternary(dbType == MySQL,
+			ansiSQLChange(`drop index argo_archived_workflows_i4 on argo_archived_workflows`),
+			ansiSQLChange(`drop index argo_archived_workflows_i4`),
+		),
+		ansiSQLChange(`create index argo_archived_workflows_i4 on argo_archived_workflows (clustername, startedat)`),
```
Comment (agilgur5) [inline persist/sqldb/migrate.go:277]: how quick does this run? quick enough to include in a patch version?

**Review by agilgur5 — COMMENTED** [review]
Huge thanks for deep-diving & analyzing this, describing it, and testing it!

**Review by agilgur5 — APPROVED** [review]

File: persist/sqldb/migrate.go (hunk 2 above)
Comment (MasonM) [reply persist/sqldb/migrate.go:277]: Very quickly. Tested after populating my DB with 100,000 rows, all associated with the same `clustername`: 

MySQL:
```
$ time go run ./hack/db migrate -d 'mysql:password@tcp/argo'
INFO[0000] Migrating database schema                     clusterName=default dbType=mysql
INFO[0000] applying database change                      change="drop index argo_archived_workflows_i4 on argo_archived_workflows" changeSchemaVersion=61
INFO[0000] applying database change                      change="create index argo_archived_workflows_i4 on argo_archived_workflows (clustername, startedat)" changeSchemaVersion=62
2024/10/26 02:24:13     Session ID:     00001
        Query:          create index argo_archived_workflows_i4 on argo_archived_workflows (clustername, startedat)
        Stack:          
                fmt.(*pp).handleMethods@/usr/local/go/src/fmt/print.go:673
                fmt.(*pp).printArg@/usr/local/go/src/fmt/print.go:756
                fmt.(*pp).doPrint@/usr/local/go/src/fmt/print.go:1208
                fmt.Append@/usr/local/go/src/fmt/print.go:289
                log.(*Logger).Print.func1@/usr/local/go/src/log/log.go:261
                log.(*Logger).output@/usr/local/go/src/log/log.go:238
                log.(*Logger).Print@/usr/local/go/src/log/log.go:260
                github.com/argoproj/argo-workflows/v3/persist/sqldb.ansiSQLChange.apply@/home/vscode/go/src/github.com/argoproj/argo-workflows/persist/sqldb/ansi_sql_change.go:11
                github.com/argoproj/argo-workflows/v3/persist/sqldb.migrate.applyChange.func1@/home/vscode/go/src/github.com/argoproj/argo-workflows/persist/sqldb/migrate.go:301
                github.com/argoproj/argo-workflows/v3/persist/sqldb.migrate.applyChange@/home/vscode/go/src/github.com/argoproj/argo-workflows/persist/sqldb/migrate.go:290
                github.com/argoproj/argo-workflows/v3/persist/sqldb.migrate.Exec@/home/vscode/go/src/github.com/argoproj/argo-workflows/persist/sqldb/migrate.go:279
                main.NewMigrateCommand.func1@/home/vscode/go/src/github.com/argoproj/argo-workflows/hack/db/main.go:50
                github.com/spf13/cobra.(*Command).execute@/home/vscode/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:985
                github.com/spf13/cobra.(*Command).ExecuteC@/home/vscode/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117
                github.com/spf13/cobra.(*Command).Execute@/home/vscode/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
                main.main@/home/vscode/go/src/github.com/argoproj/argo-workflows/hack/db/main.go:39
                runtime.main@/usr/local/go/src/runtime/proc.go:272
                runtime.goexit@/usr/local/go/src/runtime/asm_amd64.s:1700
        Rows affected:  0
        Last insert ID: 0
        Error:          upper: slow query
        Time taken:     0.95033s
        Context:        context.Background


real    0m1.758s
user    0m1.074s
sys     0m0.294s
```

PostgreSQL:
```
 $ time go run ./hack/db migrate
INFO[0000] Migrating database schema                     clusterName=default dbType=postgres
INFO[0000] applying database change                      change="drop index argo_archived_workflows_i4" changeSchemaVersion=61
INFO[0000] applying database change                      change="create index argo_archived_workflows_i4 on argo_archived_workflows (clustername, startedat)" changeSchemaVersion=62

real    0m0.868s
user    0m1.117s
sys     0m0.271s
```

**Review by isubasinghe — APPROVED** [review]
Sorry about the delay, it works  for me. I would prefer in the future  for PRs like this to be split into one for Postgres and another for MySQL please.

**Comment by MasonM** [comment]
@isubasinghe Thanks for the review! I'll make sure to split it up next time. Is there anything left to do before merging this?

**Comment by isubasinghe** [comment]
@MasonM you are good to merge.

**Comment by MasonM** [comment]
@isubasinghe I don't have write access, so I don't have permission to merge 
image

**Comment by isubasinghe** [comment]
Apologies i thought you were able to merge yourself after approval.

**Comment by htrungngx** [comment]
Hi, thanks for this fix but look like this release patch for version `3.7.x` only right ? I'm using `3.5.10` and facing same issue

**Comment by Joibel** [comment]
> Hi, thanks for this fix but look like this release patch for version `3.7.x` only right ? I'm using `3.5.10` and facing same issue

3.5 is out of support, so you'll want to update anyway.

**Pull request merged**